src/dbn_encoder.cpp

```diff
@@ -0,0 +1,266 @@
+#include "databento/dbn_encoder.hpp"
+
+#include <algorithm>  // copy
+#include <cstddef>
+#include <cstring>  // strncmp
+#include <limits>
+#include <vector>
+
+#include "databento/compat.hpp"
+#include "databento/constants.hpp"
+#include "databento/datetime.hpp"
+#include "databento/detail/zstd_stream.hpp"
+#include "databento/enums.hpp"
+#include "databento/exceptions.hpp"
+#include "databento/record.hpp"
+
+using databento::DbnEncoder;
+
+namespace {
+constexpr std::size_t kMagicSize = 4;
+constexpr std::uint32_t kZstdMagicNumber = 0xFD2FB528;
+constexpr auto kDbnPrefix = "DBN";
+constexpr std::size_t kFixedMetadataLen = 100;
+constexpr std::size_t kDatasetCstrLen = 16;
+constexpr std::size_t kReservedLen = 53;
+constexpr std::size_t kReservedLenV1 = 47;
+constexpr std::size_t kBufferCapacity = 8UL * 1024;
+
+
+
+constexpr std::uint32_t kSymbolCstrLenV1 = 22;
+constexpr std::uint64_t NULL_RECORD_COUNT = std::numeric_limits<std::uint64_t>::max();
+constexpr std::uint64_t NULL_LIMIT = 0;
+constexpr std::uint16_t NULL_SCHEMA = std::numeric_limits<std::uint16_t>::max();
+constexpr std::uint8_t NULL_STYPE = std::numeric_limits<std::uint8_t>::max();
+
+}  // namespace
+
+
+static
+std::uint32_t version_symbol_cstr_len(uint8_t version) {
+  if (version < 2) {
+    return kSymbolCstrLenV1;
+  } else {
+    return databento::kSymbolCstrLen;
+  }
+}
+
+static
+std::uint32_t calc_length(const databento::Metadata& metadata) {
+  std::uint32_t symbol_cstr_len = version_symbol_cstr_len(metadata.version);
+  std::uint32_t mapping_interval_len = sizeof(uint32_t) * 2 + symbol_cstr_len;
+  // schema_definition_length, symbols_count, partial_count, not_found_count, mappings_count
+  std::uint32_t var_len_counts_size = sizeof(uint32_t) * 5;
+
+  std::uint32_t c_str_count = metadata.symbols.size() + metadata.partial.size() + metadata.not_found.size();
+
+  std::uint32_t mappings_len = 0;
+  for (const databento::SymbolMapping& item: metadata.mappings) {
+    mappings_len += symbol_cstr_len + sizeof(uint32_t) + item.intervals.size() * mapping_interval_len;
+  }
+
+  return kFixedMetadataLen + var_len_counts_size + c_str_count * symbol_cstr_len + mappings_len;
+}
+
+
+void encode_date(std::uint32_t date, databento::IWritable& writer) {
+  // in c++ the date is already represented as a uint32, so this function doesn't really do anything compared to rust
+  std::uint8_t tmp[sizeof(date)];
```

Comment:
not while the writer expects std::uint8_t*. if its changed to expect void*, then it would work. or reinterpret_cast. static_cast from std::uint32_t* to std::uint8_t*, is not allowed (won't compile).

Reply:
Yeah, you have to `reinterpret_cast`